internal/pkg/service/stream/storage/statistics/repository/stats_move.go

```diff
@@ -0,0 +1,97 @@
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/statistics"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+// TransformFn transforms statistics value, for example to set StagingSize after the upload.
+type TransformFn func(value *statistics.Value)
+
+func (r *Repository) moveStatisticsOnSliceUpdate() {
+	r.plugins.Collection().OnSliceUpdate(func(ctx context.Context, now time.Time, original, updated *model.Slice) error {
+		fromLevel := original.State.Level()
+		toLevel := updated.State.Level()
+		if fromLevel != toLevel {
+			op.AtomicOpFromCtx(ctx).AddFrom(r.moveAll(updated.SliceKey, fromLevel, toLevel, func(value *statistics.Value) {
+				// There is actually no additional compression, when uploading slice to the staging storage
+				if toLevel == level.Staging {
+					value.StagingSize = value.CompressedSize
+				}
+			}))
+		}
+		return nil
+	})
+}
+
+// moveAll moves all statistics in the parentKey to a different storage level.
+// Returned value is sum of all slices statistics.
+func (r *Repository) moveAll(parentKey fmt.Stringer, from, to level.Level, transform ...TransformFn) *op.AtomicOp[statistics.Value] {
+	if from == to {
+		panic(errors.Errorf(`"from" and "to" storage levels are same and equal to "%s"`, to))
+	}
+
+	var sum statistics.Value
+	atomicOp := op.Atomic(r.client, &sum)
+
+	var valueKVs op.KeyValuesT[statistics.Value]
+	if k, ok := parentKey.(model.SliceKey); ok {
+		// Get
+		atomicOp.Read(func(ctx context.Context) op.Op {
+			return r.schema.InLevel(from).InSlice(k).GetKV(r.client).WithOnResult(func(kv *op.KeyValueT[statistics.Value]) {
+				if kv == nil {
+					valueKVs = nil
+				} else {
+					valueKVs = []*op.KeyValueT[statistics.Value]{kv}
+				}
```

Comment:
No, the callback can be called multiple times on some retry.

Reply:
I don't think one comment will solve it. It is a feature of all Read/Write callbacks that they are repeated in case of failure. You always have to think about it. I have no idea how to solve it better.

Reply:
Ok, I will add comment.

Reply:
https://github.com/keboola/keboola-as-code/pull/1820/commits/79ed2e24429b0f46ab29d4f4391277ab65c46ee9